core/store/src/trie/receipts_column_helper.rs

```diff
@@ -0,0 +1,164 @@
+use crate::{get, set, TrieAccess, TrieUpdate};
+use near_primitives::errors::StorageError;
+use near_primitives::receipt::{DelayedReceiptIndices, Receipt};
+use near_primitives::trie_key::TrieKey;
+
+/// Type safe access to delayed receipts queue stored in the state.
+///
+/// The struct keeps changes to indices in fields and only writes it back to
+/// state once, when `DelayedReceiptsQueue::write_back()` is called.
+#[must_use = "modified indices must be written back to state"]
+pub struct DelayedReceiptQueue {
+    indices: DelayedReceiptIndices,
+    dirty: bool,
+}
+
+/// Read-only iterator over receipt queues stored in the state trie.
+///
+/// This iterator currently only supports delayed receipts but is already
+/// written general to work with the new queues that are going to be added for
+/// congestion control.
+pub struct ReceiptIterator<'a> {
+    trie_keys: Box<dyn Iterator<Item = TrieKey>>,
+    trie: &'a dyn TrieAccess,
+}
+
+impl DelayedReceiptQueue {
+    pub fn load(trie: &dyn TrieAccess) -> Result<Self, StorageError> {
+        let indices = get(trie, &TrieKey::DelayedReceiptIndices)?.unwrap_or_default();
+        Ok(Self { indices, dirty: false })
+    }
+
+    pub fn push(
+        &mut self,
+        state_update: &mut TrieUpdate,
+        receipt: &Receipt,
+    ) -> Result<(), StorageError> {
+        let index = self.indices.next_available_index;
+        set(state_update, TrieKey::DelayedReceipt { index }, receipt);
+
+        self.indices.next_available_index = index
+            .checked_add(1)
+            .expect("Next available index for delayed receipt exceeded the integer limit");
+        self.dirty = true;
+        Ok(())
```

Comment:
Yeah, I agree. I was worried to change the behaviour too much, so I just did the exact same behaviour as we had before. But you convinced me that it's better to reduce the complexity as much as possible.

I guess I will remove the dirty flag and always write back immediately. It will be slightly less efficient, as we add an extra Vec allocation on each `set` call. But I think it's worth it.